### PR TITLE
Jrr/integrate bane

### PR DESCRIPTION
**Pull request by jrroman**
merging in bane and fixing some linter warnings. This was originally a few commits, but i couldnt clone my fork so i had to just copy my changes into my fork, hence the monolithic commit / commit msg.

### CODE REVIEW

**Comment by jrroman** [comment]
this removes the bane dependency from vendor.

File: cli/debug.go [hunk 1]
```diff
@@ -3,7 +3,7 @@ package cli
 import "log"
 
 func debuglog(message string) {
-	if global_verbose {
+	if globalVerbose {
```
Comment (GrantSeltzer) [inline cli/debug.go]: Thanks for fixing this as well!

File: generate/spec.go [hunk 2]
```diff
@@ -1,37 +1,109 @@
...
+ * http://github.com/jessfraz/bane
+ */
+type AppArmorProfileConfig struct {
+    Name         string
+    Filesystem   FsConfig
+    Network      NetConfig
+    Capabilities CapConfig
 
-// Type aliases for use from bane, temporary 
+    Imports      []string
```
Comment (GrantSeltzer) [inline generate/spec.go]: We're actually not using the Imports or InnerImports fields, you can remove them.

File: generate/spec.go [hunk 3]
```diff
@@ -1,37 +1,109 @@
...
+    InnerImports []string
+}
+
+// FsConfig defines the filesystem options for a profile.
+type FsConfig struct {
```
Comment (GrantSeltzer) [inline generate/spec.go]: FsConfig -> Filesystem

File: generate/spec.go [hunk 4]
```diff
@@ -1,37 +1,109 @@
...
+    WritablePaths   []string
+    AllowExec       []string
+    DenyExec        []string
+}
+
+// NetConfig defines the network options for a profile.
+// For example you probably don't need NetworkRaw if your
+// application doesn't `ping`.
+// Currently limited to AppArmor 2.3-2.6 rules.
+type NetConfig struct {
```
Comment (GrantSeltzer) [inline generate/spec.go]: NetConfig -> Network

File: generate/spec.go [hunk 5]
```diff
@@ -1,37 +1,109 @@
...
+    Raw       bool
+    Packet    bool
+    Protocols []string
+}
+
+// CapConfig defines the allowed or denied kernel capabilities
+// for a profile.
+type CapConfig struct {
```
Comment (GrantSeltzer) [inline generate/spec.go]: CapConfig -> Capabilities

**Comment by GrantSeltzer** [comment]
Just a couple of changes I requested but this is great otherwise. Thanks a lot!

**Comment by GrantSeltzer** [comment]
Ah, also you can remove the dependency from vendor, looks like you just deleted the VERSION file. 

You can use [Dep](https://github.com/golang/dep) to do this easily.

Once you have that installed (`go get -u github.com/golang/dep`) you can just run `dep ensure -v` at the base of the repo. it'll see that bane is no longer a dependency and remove it.

File: generate/spec.go [hunk 6]
```diff
@@ -1,37 +1,109 @@
...
+    Allow []string `toml:"Allow,omitempty"`
+    Trap  []string `toml:"Trap,omitempty"`
+    Trace []string `toml:"Trace,omitempty"`
+    Kill  []string `toml:"Kill,omitempty"`
+    Errno []string `toml:"Errno,omitempty"`
 }
 
+/**
+ * Originally Written by Jess Frazelle
+ * http://github.com/jessfraz/bane
```
Comment (GrantSeltzer) [inline generate/spec.go]: I like giving credit but i'll do that in a contributors file or something, don't need this in the codebase.

**Comment by jrroman** [comment]
Awesome not a problem i will address these issues thanks!

**Comment by jrroman** [comment]
okay i have addressed the above issues let me know if there was anything missed.

File: generate/spec.go [hunk 7]
```diff
@@ -1,37 +1,93 @@
...
-type Capabilities          = bane.CapConfig
+    if err := compiled.Execute(out, profile); err != nil {
+        return err
+    }
+
+    return nil
+}
+
+// check if the tunables/global exist
+func tunablesExists(name string) bool {
```
Comment (GrantSeltzer) [inline generate/spec.go]: I guess we won't need this

File: generate/spec.go [hunk 8]
```diff
@@ -1,37 +1,93 @@
...
+    _, err := os.Stat(path.Join("/etc/apparmor.d/tunables", name))
+    return err == nil
+}
+
+// check if abstractions/base exist
+func abstractionsExists(name string) bool {
```
Comment (GrantSeltzer) [inline generate/spec.go]: or this :)

**Comment by GrantSeltzer** [comment]
Last thing, then we can merge. Thanks!

**Comment by jrroman** [comment]
My oversight sorry about that, changes have been made and committed.

**Comment by GrantSeltzer** [comment]
No worries @jrroman, really appreciate the contrib! 

You did however forget to remove the unused imports which causes compile errors :)
```
# github.com/GrantSeltzer/karn/generate
generate/spec.go:4:5: imported and not used: "os"
generate/spec.go:6:5: imported and not used: "path"
```

Looks like you're not using an IDE that does stuff like that for you. My personal favorite is [VSCode](https://code.visualstudio.com/) which I highly recommend. The Go plugin will auto remove unused imports among many other really convenient features like 'Find All References' or 'Peek Definition'

**Comment by jrroman** [comment]
i will look into vscode for sure, I removed those unnecessary imports, i also fixed the outstanding spec.go lint errors.

**Comment by jrroman** [comment]
I do know that if err := func(); err != nil { return err } is common syntax, but the linter was complaining about redundant "if's" so that is why i changed the style.

File: generate/spec.go [hunk 9]
```diff
@@ -73,7 +71,8 @@ func (profile *AppArmorProfileConfig) Generate(out io.Writer) error {
         return err
     }
 
-    if err := compiled.Execute(out, profile); err != nil {
+    err = compiled.Execute(out, profile)
+    if err != nil {
```
Comment (GrantSeltzer) [inline generate/spec.go]: Yea I prefer this style

**Comment by GrantSeltzer** [comment]
Awesome, thanks a ton @jrroman !

**Comment by jrroman** [comment]
Not a problem!

**Comment by jrroman** [comment]
@GrantSeltzer Awesome man i really appreciate that, I forgot that i had switched the email that my account was linked to hence the missing data! Thank you my friend!

**Comment by GrantSeltzer** [comment]
Sweet :)

**Pull request merged**